source/extensions/tracers/zipkin/zipkin_core_types.cc

```diff
@@ -219,6 +238,41 @@ const std::string Span::toJson() {
   return json_string;
 }
 
+const zipkin::proto3::Span Span::toProtoSpan() const {
+  zipkin::proto3::Span span;
+  span.set_trace_id(traceIdAsByteString());
+  span.set_parent_id(parentIdAsByteString());
+  span.set_id(idAsByteString());
+  span.set_name(name_);
+
+  if (timestamp_) {
+    span.set_timestamp(timestamp_.value());
+  }
+
+  if (duration_) {
+    span.set_duration(duration_.value());
+  }
+
+  for (const auto& annotation : annotations_) {
+    if (annotation.isSetEndpoint()) {
+      if (annotation.value() == ZipkinCoreConstants::get().CLIENT_SEND) {
+        span.set_kind(zipkin::proto3::Span::CLIENT);
+        span.mutable_local_endpoint()->MergeFrom(annotation.endpoint().toProtoEndpoint());
+      } else if (annotation.value() == ZipkinCoreConstants::get().SERVER_RECV) {
+        span.set_kind(zipkin::proto3::Span::SERVER);
+        span.mutable_remote_endpoint()->MergeFrom(annotation.endpoint().toProtoEndpoint());
+      }
+    }
+  }
+
+  auto& tags = *span.mutable_tags();
+  for (const auto& binary_annotation : binary_annotations_) {
+    tags[binary_annotation.key()] = binary_annotation.value();
```

Comment:
careful. boolean annotations of "ca" or "sa" include the remote endpoint. this should check if binary annotation is type string, unless the only allowed type is string.